feat(product-import): Set product state when not specified

michael-s-grant opened the pull request:
#### Summary

Set state to New for REWE Digital Marketplace partner products if state is not specified.  Fixes #125/[PIN-1887](https://jira.rewe-digital.com/browse/PIN-1887).*

#### Description
This work emerged out of the [addition of states](https://confluence.rewe-digital.com/pages/viewpage.action?pageId=364020405)* to the REWE Marketplace product model.  When new products are imported, the state is now set to `New` if it is not specified.

The pull request also includes a bit of refactoring.

∗ Sorry, open-source contributors, this link will not be accessible from outside of commercetools and REWE Digital.

- Tests
    - [X] Unit
    - [X] Integration
    - [ ] Acceptance
- [ ] Documentation

Commit:
feat(productImport): Set state to New for partner products if state is not specified

Set state to New for REWE Digital Marketplace partner products if state is not specified.
Also, refactor timeout in integration.spec.coffee into a common constant, and increase its value.

Closes: #125

Commit:
refactor(productImport): refactor to extract out duplicated code, and for shorter methods

Refactor to extract duplicated code out from _prepareUpdatedProduct and
_prepareNewProduct by creating a new method _assignCommonIds, and for
consistency of variable names between these two methods.  Also to shorten
_resolveReference by creating a new method _resolveStateReferenceForProductWithoutOne.

Commit:
test(productImport): make tests clean up states; also add a comment

Make tests clean up states; also add a comment to product-import.coffee
explaining that promises are being cached to avoid unnecessary calls to CTP.

junajan commented on lib/product-import.coffee:
```diff
@@ -602,26 +617,55 @@ class ProductImport
+
+  _resolveStateReferenceForProductWithoutOne: (service, refKey, predicate, resolve, reject) =>
+    if @_cache[refKey][predicate]
+      if (typeof @_cache[refKey][predicate] is 'object' and @_cache[refKey][predicate].constructor.name is 'Promise')
```
You can simplify this with function `_.memoize` from lodash.

junajan commented on test/integration/product-import.spec.coffee:
```diff
@@ -18,7 +18,7 @@ sampleProductTypeForProduct =
 bigProductType =
   name: 'bigProductType'
-  description: 'test big poroductType description'
+  description: 'test big productType description'
```
👍

junajan commented on test/integration/product-import.spec.coffee:
```diff
@@ -179,6 +196,9 @@ describe 'Product Importer integration tests', ->
       .then => cleanup(logger, @client.customerGroups, @customerGroup.id)
       .then => cleanup(logger, @client.channels, @channel.id)
       .then => cleanup(logger, @client.categories, @category.id)
+      .then => cleanup(logger, @client.states, @newState.id)
```
here we could use something like `cleanAll` which would clean all from the given resource (eg states).

junajan commented on test/integration/product-import.spec.coffee:
```diff
@@ -122,6 +122,21 @@ sampleCustomerGroup =
+  initial: true
+
+extantState =
+  key: "AlsoNewButIWantToCheckPopulatingWithANonDefaultState"
```
Is this long key necessary? IMHO shorter is better

michael-s-grant replied:
I can certainly move the description from thr key into a comment.

junajan commented on test/integration/product-import.spec.coffee:
```diff
@@ -216,6 +236,98 @@ describe 'Product Importer integration tests', ->
+        done()
+
+  it 'should process updates for products where the update does not specify the state', (done) ->
+    productDraft = _.deepClone(createProduct()[0])
```
why `deepClone`?

michael-s-grant replied:
Just copying what was there already in other tests.

junajan reviewed (changes requested):
Hello @michael-s-grant .. It looks pretty good, I added just a few comments.

butenkor commented on lib/product-import.coffee:
```diff
@@ -475,32 +489,33 @@ class ProductImport
+      statePredicate = "id=\"#{productToProcess.state.id}\" and type=\"ProductState\""
+    else
+      stateRef = ''
+      statePredicate = 'key="New" and type="ProductState"'
```
This is generalised implementation. Lets not make it customer specific one. If default state has to be used than it has to be set in the product input draft already (preferred variant). Another possibility would be to introduce new optional configuration parameter with following name and value: `defaultProductStateKey`: `new`.

Commit:
refactor(productImport): extract duplicated code into a new method

Extract code duplicated between _prepareNewProduct and _prepareUpdateProduct
into a new method, _resolveStateReference.

Commit:
refactor(productImport): shorten key of test state

extantState had the key AlsoNewButIWantToCheckPopulatingWithANonDefaultState.
Shorter keys are, so I am told, better, so I have renamed it AlsoNew and put
the rest of the description into a comment.

michael-s-grant commented on test/integration.spec.coffee:
```diff
@@ -18,6 +18,13 @@ sampleTaxCategory = require '../samples/sample-tax-category.json'
+  type: "ProductState"
+  initial: true
+
+TEST_TIMEOUT = 15000
```
Well, I had set it in just the places where the hardwired value was beforehand, but I can certainly follow your suggestion, doing it the same way it's done in integration/product-import.spec.coffee.  It needs to go up anyway, I'm still getting tests intermittently timing out.

Commit:
refactor(productImport): changed implementation of and lengthened test timeout

Increased test timeout from 15" to 30", as with the latter I was still seeing
tests timing out intermittently; and changed it from an explicit timeout to
setting Jasmine's default timeout interval, the same way it's done in
integration/product-import.spec.coffee.

michael-s-grant commented on lib/product-import.coffee:
```diff
@@ -448,23 +448,37 @@ class ProductImport
+    if (productToProcess.state)
+      stateRef = productToProcess.state
+      statePredicate = "id=\"#{productToProcess.state.id}\" and type=\"ProductState\""
+    else
```
As @butenkor suggested elsewhere, I think it would be good to implement a config option so we can insist on states for Marktplatz usage without foisting them on other users.  (Now I've just got to figure out how to do this via the way the sphere-product-importer is called for us via, apparently, the sphere-node-product-csv-sync.)

daern91 commented:
Hi @michael-s-grant, is this PR still alive and waiting for approval?

michael-s-grant commented:
@daern91, in response to your question (which I don't seem to be able to answer directly), no, this work got superseded by that on the sphere-node-product-csv-sync project.  If you think it might come in useful, you could still merge this pull request, but I am no longer working on it.

daern91 commented:
Got it, @butenkor and @junajan, what do you think?

junajan commented:
Hello @daern91 .. from my point of view:
 - this module is being used by other projects as an external dependency
 - it is also used by `sphere-node-cli` which takes products in json and send it to this module

In the future, we want to get rid of `sphere-node-product-csv-sync` we will probably create a new `monorepo` package which maps products in CSV to JSON and then send it to this module so here we should have the same functionality as it is in `sphere-node-product-csv-sync`.

So I would prefer to add the same change also here and get rid of csv-sync as fast as possible so we won't have to maintain two projects instead of just one

daern91 reviewed (commented):
This brings some old memories :)

michael-s-grant commented:
Closing pull request probably long overtaken by subsequent developments.
